Refactoring the options and configuration framework (#3409)
informalsystems/apalache: docs/src/apalache/config.md

# Apalache configuration

Apalache configuration files use **strict JSON**. The supported filenames are
`.apalache.json` for a project file and `$HOME/.tlaplus/apalache.json` for a user-wide file.

## Loading and precedence

> Values are applied in this order, from highest to lowest precedence:

Reviewer: I think this is contradictory.

If they are applied in this order, the later ones would take presendence, but you list them from high to low precedence?

@copilot suggest better wording

1. command-line arguments;
2. environment variables used by command-line options;
3. the file passed with `--config-file`, or the nearest `.apalache.json` found by searching the current directory and
   its parents;

> 4. `$HOME/.tlaplus/apalache.json`;

Reviewer: this suggests `$HOME/.tlaplus/apalache.json` is applied even `.apalache.json` is found or `--config-file` is given?

Author: Yeah. This whole policy is confusing. Let me check how it is actually done. I would also rather prefer not merging configuration files. It's easy to merge them, but this behavior is confusing. Shall we agree on merging CLI options with the config files (CLI options take priority over the config files), but using only a single config files, whichever wins?

Reviewer: Yes, this sounds good to me 👍🏻

Author: Ugh. This whole discussion surface a potential security issue. The current implementation allows the user to specify inputs and outputs in the configurations. Also, fixing that.

Author: Fixed all of the things discussed here.

5. built-in defaults.

> An explicit `--config-file` disables the search for a local file, but not the global fallback. Lists and ordinary values

Reviewer: Why? I cannot think of a case where I want to give an explicit `--config-file` and still use a global fallback.

replace lower-precedence values.

> `checker.tuning` objects are merged by key.

Reviewer: out of place?

The old `.apalache.cfg` and `$HOME/.tlaplus/apalache.cfg` application configuration filenames are rejected, even when
they contain valid JSON or a
`.json` file exists beside them. Rename them to `.apalache.json` and
`$HOME/.tlaplus/apalache.json`, respectively. Explicit `--config-file` paths ending in `.cfg` are rejected as well.

> HOCON syntax is no longer supported.

Reviewer: This entire paragraph needs to be rewritten with the reader in mind.

```
Apalache previously supported `.cfg` configuration files in HOCON syntax.
These are no longer supported.
...
```

This does not affect TLC configuration files passed with `--config`, which files conventionally use the `.cfg`

> extension.

Reviewer: ```suggestion
This does not affect TLC configuration files passed with `--config`, that conventionally use the `.cfg` extension.
```

**rejected**: Comments, substitutions such as `${PWD}`, unquoted keys, `=`, trailing commas, duplicate keys, and
trailing documents. **Unknown keys are rejected** in every group; for example, `checker.discardDisabled`
is an error.

`source` and `output` are top-level values. The former `input` object and object-valued `output` section are rejected.

A leading `~` or `~/` in a path expands to the user's home directory. Other environment-variable expansion is not
performed.

This is a complete, copyable example:

```json
{
   "out-dir": "./_apalache-out",
   "debug": false,
   "smtprof": false,
   "write-intermediate": false,
   "profiling": false,
   "features": [],
   "source": "./Spec.tla",
   "output": "./Parsed.tla",
   "checker": {
      "algo": "incremental",
      "discard-disabled": true,
      "length": 10,
      "max-error": 1,
      "timeout-smt": 0,
      "no-deadlock": false,
      "smt-solver": "z3",
      "smt-encoding": "oopsla19",
      "tuning": {}
   },
   "typechecker": {
      "infer-poly": true
   },
   "server": {
      "port": 8822,
      "server-type": "checker"
   }
}
```

## Supported keys

> An absent default of “none” means that the value is genuinely optional.

Reviewer: ```suggestion
In the table below, a default of "none" means that the value is optional.
```

| top level     | `out-dir`            | Base directory for generated run directories.                | path string                  | `./_apalache-out`                           |
|               | `run-dir`            | Also write this run's output directly to this directory.     | path string                  | none                                        |
|               | `debug`              | Enable detailed logging and a configuration snapshot.        | boolean                      | `false`                                     |
|               | `smtprof`            | Write an SMT constraint profile.                             | boolean                      | `false`                                     |
|               | `write-intermediate` | Save intermediate representations produced by passes.        | boolean                      | `false`                                     |
|               | `profiling`          | Write profiling data for transformation rules.               | boolean                      | `false`                                     |

> |               | `features`           | Enable experimental language or checker features.            | array of strings             | `[]`; names shown by CLI help               |

Reviewer: Didn't know we have this 😂 

afaict, it's only used for `rows`? Should we remove it in a future PR?

Author: Totally. This is also on my refactoring list https://github.com/apalache-mc/apalache/issues/3405

Top-level `command` and `config-file` can appear in JSON exchanged with the RPC API and in diagnostic snapshots. Normal
configuration files should not set them; the selected command and `--config-file` provide those values.

A file source is normally just a path string. RPC callers may provide an in-memory source:

```json
{
   "source": {
      "kind": "string",
      "content": "---- MODULE M ----\n====",
      "aux": [],
      "format": "tla"
   }
}
```

The source formats are `tla`, `json`, `itf`, and `qnt`. A file with a nonstandard or ambiguous extension can use a
source object with
`"kind": "file"`, `"path": "..."`, and an explicit `"format"`.

## TLC configuration precedence

For `init`, `next`, `inv`, `temporal`, and deadlock checking, an application configuration value overrides the TLC file.
Otherwise the TLC value is used, then the application default. `CHECK_DEADLOCK FALSE` is equivalent to
`"no-deadlock": true`.

## Deprecated JSON aliases

The following old spellings remain accepted with warnings:

> | Deprecated                | Canonical                |

Reviewer: since we anyhow break compatiblity, should we just retire them now?

Author: I wanted to keep them for compatibility with quint, but let me check, whether we actually need them

Reviewer: Are these still present? 
We could at least make a follow-up issue.

|---------------------------|--------------------------|
| `checker.timeout-smt-sec` | `checker.timeout-smt`    |
| `checker.no-deadlocks`    | `checker.no-deadlock`    |
| `checker.temporal-props`  | `checker.temporal`       |
| `typechecker.inferpoly`   | `typechecker.infer-poly` |

> Old object-shaped enum and source values using a `type` field are also read with a warning. Setting a canonical key and

Reviewer: I have no idea what this means.

Probably better to skip it in the docs.
